### DISCUSSION TITLE
Adding variables to Explore screen

### DISCUSSION THREAD

**Discussion by dave08** [grafana/grafana#52013] [Explore]
Often enough there are a bunch of places that need to be changed at once in promql queries (and other types too), especially when using multiple metrics with similar lablels. It would be nice to be able to configure variables in explore just like in dashboards to simplify testing out different queries.

It could also simplify exploring queries from dashboards -- instead of injecting the variables in them, just copy them over to the explore view with the variables...

@gelicia

**Comment by gelicia** [comment]
This is related to https://github.com/grafana/grafana/discussions/45673 involving moving variables from Dashboards to explore. Thanks for creating this feature request!

**Comment by gelicia** [comment]
Some other parts of this feature
- MVP is just a text field for a key/value match, and the key can be put into queries. Changing the value changes the value in all areas.
- Next step is making variables as powerful as dashboard variables. "What if, other than "free text" variables, I would also have query-based variables? I would type a query to extract all possible $namespace values and then a simple dropdown menu would allow me to switch between namespaces."
- Have variables in the URL, so sharing an explore URL shares those variables (potential for too long of a URL? should we put this somewhere else? how to handle query variables?)

**Comment by dave08** [comment]
The text field option is currently the most useful one, instead of just plugging in the values from the dashboard into the queries, copy them over as variables when going to the explore view...

The other options could get a bit complicated as long as there's no server-side history and favorite query feature... if there would have been that, then there could just be a short url to such a view. But if not, how much would one want to work to set up a temporary query?

The option of simple text variables are very powerful though, since they avoid repetition in writing those temporary queries and allow changing them all in one shot simply by changing the value of the text variable.

I'm not saying that having more advanced variables wouldn't be useful, I'm just saying that they might be implemented a bit differently than the dashboard ones, since the purpose of the explore view is very different.

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
